---
title: "Human muscle transcriptomics (file 5)"
output:
  html_notebook: default
  word_document: default
editor_options:
  chunk_output_type: console
  always_allow_html: true
---

```{r setup4, include = FALSE}
rm (list = ls ())
suppressPackageStartupMessages ({ 
  library (data.table)
  library (limma)
  library (edgeR)
  library (dplyr)
  library (WGCNA)
  library (biomaRt)
  library (flashClust)  
  library (ggplot2)
  library (scales) 
  library (dplyr)
  library (plotly)
  library (knitr)
  library (scales)
  library (cowplot)
  library (tidyr)
  library (grid)
  library (ggpubr)
  })
opts_chunk$set (eval = TRUE, echo = TRUE, tidy = TRUE, highlight = TRUE, dev = c ('png'), fig.width = 10, fig.height = 8,  fig.path = "figures/")

BiocStyle::markdown ()
options (scipen = 999)

# RNA isolation protocol
WithKit <- c ("MD06_GRA_Sam25", "MD07_GRA_Sam23", "MD08_GRA_Sam65", 
              "MD10_GRA_Sam106", "MD12_GRA_Sam139")
WithoutKit <- c ("MD06_GRA_Sam148", "MD07_GRA_Sam73_RedoSam15", "MD08_GRA_Sam152_RedoSam39", 
                 "MD10_GRA_Sam36", "MD12_GRA_Sam116")
# Color and pch
pch.muscle <-  c ("GAL" = 7, "GRA" = 1, "REF" = 4, "SEM" = 2, "SED" = 3, "VAL" = 5, "VAM" = 6)
col.muscle <- c ("GAL" = "#999999", "GRA" = "#0099FF", "REF" = "#FF9999", 
                 "SED" = "#0033FF", "SEM" = "#0066FF", "VAL" = "#FF6666", 
                 "VAM" = "#FF3333")
col.batch <- c ("1" = "#e6194B", "2" = "#3cb44b", "3" = "#4363d8", "4" = "#911eb4", "5" = "black")
col.individual <- c ("MD06" = "#A6CEE3", "MD07" = "#1F78B4", "MD08" = "#B2DF8A", 
                     "MD10" = "#33A02C", "MD11" = "#FB9A99", "MD12" = "#E31A1C", 
                     "MD13" = "#FDBF6F", "MD14" = "#FF7F00", "MD15" = "#66C2A5", 
                     "MD16" = "#FC8D62", "MD17" = "#8DA0CB", "MD18" = "#E78AC3", 
                     "MD19" = "#A6D854", "MD20" = "#FFD92F", "MD21" = "#E5C494", 
                     "MD22" = "#B3B3B3", "MD26" = "#1B9E77", "MD27" = "#D95F02", 
                     "MD28" = "#7570B3", "MD31" = "#E7298A")
col.lane <- c ("L003.1" = "lightblue", "L004.1" = "darkblue", "L002.2" = "pink", "L004.2" = "red")
col.RNA <- c ("WithKit" = "#bfef45", "WithOutKit" = "#f58231")
col.Hos <- c ("EMC" = "green", "HMC" = "orange")

load ("Outputs/CorrectedDatasetToAnalysis.RData")

# Colors and labels for plots
MusLabPlot <- gsub ("MD[0-9][0-9]_|_Sam.*", "", colnames (D_Raw))
MusColPlot <- col.muscle [match (gsub ("MD[0-9][0-9]_|_Sam.*", "", colnames (D_Raw)), 
                                  names (col.muscle))]
IndLabPlot <- gsub ("_.*", "", colnames (D_Raw))
IndColPlot <- col.individual [match (gsub ("_.*", "", colnames (D_Raw)), 
                                  names (col.individual))]
BatchLabPlot <- D_Raw$samples$batch [match (colnames (D_Raw), rownames (D_Raw$samples))]
BatchColPlot <- col.batch [match (D_Raw$samples$batch [match (colnames (D_Raw), 
                                                          rownames (D_Raw$samples))], 
                                names (col.batch))]
RNALabPlot <- D_Raw$samples$IsolationRNA [match (colnames (D_Raw), rownames (D_Raw$samples))]
RNAColPlot <- col.RNA [match (D_Raw$samples$IsolationRNA [match (colnames (D_Raw), 
                                                             rownames (D_Raw$samples))], 
                                names (col.RNA))]
LaneLabPlot <- D_Raw$samples$lane [match (colnames (D_Raw), rownames (D_Raw$samples))]
LaneColPlot <- col.lane [match (D_Raw$samples$lane [match (colnames (D_Raw), 
                                                       rownames (D_Raw$samples))], 
                                names (col.lane))]
HosLabPlot <- D_Raw$samples$Hospital [match (colnames (D_Raw), rownames (D_Raw$samples))]
HosColPlot <- col.Hos [match (D_Raw$samples$Hospital [match (colnames (D_Raw), 
                                                       rownames (D_Raw$samples))], 
                                names (col.Hos))]
```

## B. Differential expression analysis
We used the linear mixed models considering individual as a random effect, and muscle and batch as fixed effects.


The lmer function form lmerTest R package is used. And to find if there is any diffrence between muscles or batches the anova is performed. For the pairwise comparison we used Least-Squares Means (lsmeants R function)

```{r DEA, echo = FALSE, warning = FALSE}
if (file.exists ("Outputs/DE/DEtest_CorrectedBatchMus.RData")) {
  load ("Outputs/DE/DEtest_NotCorrected.RData")
  load ("Outputs/DE/DEtest_CorrectedBatch.RData")
  load ("Outputs/DE/DEtest_CorrectedBatchMus.RData")} else {
    for (Dfile in ls (pattern = "D_")){
      D <- get (Dfile)
      # Creating a design matrix 
      if (Dfile == "D_Raw") {
        design <- model.matrix (~ group + individual + batch , data = D$samples) } else {
          design <- model.matrix (~ group + individual , data = D$samples)
        }
      colnames (design) <- gsub ("group|individual", "", colnames (design))
      # mean-variance trend was removed using the voom
      v <- voom (D, design, plot = TRUE)
      DEtest <- matrix (data = NA, nrow = nrow (v$E), ncol = 140)
      rownames (DEtest) <- rownames (v$E)
      AllPossPairs <- combn (1 : length (names(col.muscle)), 2, simplify = FALSE)
      colnames (DEtest) [112:140] <- c ("AveExpr", paste0 ("AveExpr", sort (names (col.muscle))),                              paste0 ("Log2FC_", unlist (lapply (AllPossPairs, function (x) 
      paste (names (col.muscle[x]), collapse = "_"))))) 
      if (Dfile == "D_Raw") { 
        DEbatch <- matrix (data = NA, nrow = nrow (v$E), ncol = 16) 
        rownames (DEbatch) <- rownames (v$E)
        DEfixef <-  matrix (data = NA, nrow = nrow (v$E), ncol = 11)} else {
          DEfixef <-  matrix (data = NA, nrow = nrow (v$E), ncol = 7)
        }
      
      rownames (DEfixef) <- rownames (v$E)
      # fit linear-mixed model for each gene (including individual as random and all factors as fixed effects)
      # RIN is excluded as we have missing value for 12 samples
      for (gene in 1:nrow (v$E)) {
        print (gene)
        # Data
        DE.Data <- cbind (v$E [gene, ], v$weights [gene, ], v$targets)
        colnames (DE.Data)[1:2]<- c ("values", "weights")
        # fit LMM 
        if (Dfile == "D_Raw") {
          MM <- lmerTest::lmer (values ~ group + batch + (1|individual), weights = weights, data = DE.Data) 
        } else {
            MM <- lmerTest::lmer (values ~ group + (1|individual), weights = weights, data = DE.Data) }
        # Muscle
        # Pairwise comparison 
        PairWiseComp <- as.data.frame (lsmeans::lsmeans (MM, pairwise ~ group)$contrast)
        PairWiseComp <- reshape2::melt (PairWiseComp, id = "contrast"); 
        PairWiseComp <- PairWiseComp [order (PairWiseComp$contrast), ]
        DEtest [gene, 1:111] <- c (unlist (anova (MM)["group",]), PairWiseComp$value)
        # Extract the fixed-effect coefficients using fixef()
        DEfixef [gene, ] <- lme4::fixef (MM)
        # Average expression
        DEtest [gene, 112] <- mean (DE.Data$values)
        # Average for muscle
        meanMus <- data.frame (DE.Data %>% group_by (group) %>% summarise (mean (values)))
        meanMus <- meanMus [sort (meanMus$group) ,]
        DEtest [gene , c (113:119)] <- meanMus [ sort (meanMus$group), 2]
        # Log2FC
        logFC <- c()
        for (i in 1:length(AllPossPairs)){
          p <- AllPossPairs [[i]]
          logFC [i] <- meanMus [p[1],2] - meanMus [p[2],2]
        }
        DEtest[gene , c (120:140)] <- logFC
        if (Dfile == "D_Raw") {
          # Batch
          PairWiseCompB <- as.data.frame (lsmeans::lsmeans (MM, pairwise ~ batch)$contrast)
          PairWiseCompB <- reshape2::melt (PairWiseCompB[,c(1,6)], id = "contrast"); 
          PairWiseCompB <- PairWiseCompB [order (PairWiseCompB$contrast), ]
          DEbatch [gene, ] <- c(unlist (anova (MM)["batch",]), PairWiseCompB$value)
        }
      }
      DEtest <- as.data.frame (DEtest)
      colnames (DEtest) [1:111] <- c (colnames (anova(MM)["group",]), 
                                      paste ( PairWiseComp$contrast , PairWiseComp$variable))
      if (Dfile == "D_Raw") { 
        DEbatch <- as.data.frame (DEbatch)  
        colnames (DEbatch) <- c (colnames (anova(MM)["batch",]),
                                 paste ( PairWiseCompB$contrast , PairWiseCompB$variable)) 
      }
      DEfixef <- as.data.frame (DEfixef)
      colnames (DEfixef) <- names (lme4::fixef (MM))
      # I see to warning messages:
      ## (1) boundary (singular) fit: see ?isSingular
      # We have this warning when the Std.Dev. for random effect is equal to zero
      # (https://bbolker.github.io/mixedmodels-misc/glmmFAQ.html#singular-models-random-effect-variances-estimated-as-zero-or-correlations-estimated-as---1)
      ## (2) In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :
      ##   Model failed to converge with max|grad| = 0.0033159 (tol = 0.002, component 1)
      # We have this warning when the random effect (individual) is explaining everything and leaving novariation at all for either any fixed effect or the residual variance.
      # (https://stackoverflow.com/questions/53034261/warning-lme4-model-failed-to-converge-with-maxgrad)

      DEtest$`FDR(>F)` <- p.adjust (DEtest[,"Pr(>F)"], method = "fdr")
      if (Dfile == "D_Raw") { DEbatch$`FDR(>F)` <- p.adjust (DEbatch[,"Pr(>F)"], method = "fdr") }
      
      # Saving the dataset for the analysis
      assign (gsub ("D", "DEtest", Dfile), DEtest)
      assign (gsub ("D", "DEfixef", Dfile), DEfixef)
      assign (gsub ("D", "v", Dfile), v)

      if (Dfile == "D_Raw") {
        assign (gsub ("D", "DEbatch", Dfile), DEbatch)
        save (v_Raw, DEtest_Raw, DEbatch_Raw, DEfixef_Raw, file = "Outputs/DE/DEtest_NotCorrected.RData")
        }
      if (Dfile == "D_Batch") {
        save (v_Batch, DEtest_Batch, DEfixef_Batch, file = "Outputs/DE/DEtest_CorrectedBatch.RData")
      }
      
      if (Dfile == "D_BatchMus") {
        save (v_BatchMus, DEtest_BatchMus, DEfixef_BatchMus, file = "Outputs/DE/DEtest_CorrectedBatchMus.RData")

      }
    }
  }
```

### The batch effect
The anova is used to find the genes that show different expression levels considering the batch as covariate.
```{r DEbatch, echo = FALSE, warning = FALSE, message = FALSE}
# Subseting the deferentially expressed genes based on batch
DEtestSigBatch <- DEbatch_Raw %>% tibble::rownames_to_column () %>%
  filter (`FDR(>F)` < 0.05) %>%
  tibble::column_to_rownames ("rowname")

print ("Number of genes with different expression levels in different batches") 
print (nrow (DEtestSigBatch)) 
```

We use lsmeans (least square means) to have pairwise comparison between batches (batch is the the covariate). Least square means are means for groups that are adjusted for means of other factors in the model. 

The below plot shows the number of genes (FDR < 0.05) with different expression levels in the pairwise comparisons of all the batches. The color intensity indicates the number of genes with different expression levels between batches (the darker the color, the higher the number).

```{r DEbatchTable, echo = FALSE, warning = FALSE, message = FALSE}
# Counting the number of deferentially expressed genes in each pairwise comparisons of batches
ColPVAl <- colnames (DEbatch_Raw)[grep (" p.value" , colnames (DEbatch_Raw))]
BatchDissSim <- matrix (0, nrow = 5, ncol = 5) 
colnames (BatchDissSim) <- rownames (BatchDissSim) <- names (col.batch)
for (c in ColPVAl){
BatchDissSim [gsub (".*- | p.*", "", c), gsub (" -.*", "", c)] <- BatchDissSim [gsub (" -.*", "", c), gsub (".*- | p.*", "", c)] <- table (DEtestSigBatch [c]< 0.05)["TRUE"]
}
BatchDissSim [ is.na (BatchDissSim) ] = 0
colnames (BatchDissSim) <- rownames (BatchDissSim) <- paste ("Batch", names (col.batch))

print (knitr::kable (BatchDissSim, align = "c"))
```

```{r DEbatchPlot, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 5, fig.width = 5}
gplots::heatmap.2 (BatchDissSim, col = colorRampPalette (c ("white", "black")), 
            trace = "none", margins = c (7, 7), breaks = 100)
```

Batch 5 is the most different one. This result is in agreement with our unsupervised clustering of the samples.

### The muscle effect
The anova is used to find the genes that show different expression levels considering the muscle type as covariate.
```{r DEmuscle, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 5, fig.width = 5}
# Subsetting the differentially expressed genes
for (DEfile in ls (pattern = "DEtest_")){
  print (DEfile)
  DEtest <- get (DEfile)
  DEtestSig <- DEtest %>% tibble::rownames_to_column () %>%
    filter (`FDR(>F)` < 0.05) %>%
    arrange (`FDR(>F)`) %>%
    tibble::column_to_rownames ("rowname") 
  print ("Number of DE genes") 
  print (nrow (DEtestSig))
  assign (gsub ("_", "Sig_", DEfile), DEtestSig)
  rm (DEfile, DEtest, DEtestSig)
}
```

We use lsmeans to have pairwise comparison between muscles (muscle type is the the covariate).
There are 21 different pairwise comparisons. In the below table, the first column is the number of pairwise comparison that the gene is deferentially expressed and the second column is the number of genes  

There are "Raw: 241, BatchMus: 294 and Batch: 257" genes with significant difference in expression levels across the muscles (anova) but with no significant pairwise comparison (lsmeans).
```{r DEmusclePairwise, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 5, fig.width = 5}
for (DEfile in ls (pattern = "DEtestSig_")){
  DEtestSig <- get (DEfile)
  print (DEfile)
  print (knitr::kable (data.frame (table (rowSums (DEtestSig [ , grep ("p.val", colnames (DEtestSig))]< 0.05)))))
  rm (DEfile, DEtestSig)
}
```

The below plot shows the number of deferentially expressed genes (FDR < 0.05) in the pairwise comparisons of all the muscles.
The color intensity indicates the number of DE genes (the darker the color, the higher the number).

```{r DEmuscleTable, echo = FALSE, warning = FALSE, message = FALSE}
for (DEfile in ls (pattern = "DEtest_")){
  print (DEfile)
  DEtest <- get (DEfile)
  DEtestSig <- get (gsub ("_", "Sig_", DEfile))
  # Counting the number of differentially expressed genes in each pairwise comparisons
  ColPVAl <- colnames (DEtest)[grep (" p.value" , colnames (DEtest))]
  MusDissSim <- matrix (0, nrow = 7, ncol = 7) 
  colnames (MusDissSim) <- rownames (MusDissSim) <- names (col.muscle) 
  for (c in ColPVAl){
    MusDissSim [gsub (".*- | p.*", "", c), gsub (" -.*", "", c)] <- MusDissSim [gsub (" -.*", "", c), gsub (".*- | p.*", "", c)] <- table (DEtestSig [c]< 0.05)["TRUE"]
    }
  MusDissSim [ is.na (MusDissSim) ] = 0
  print (knitr::kable (MusDissSim, align = "c"))
  assign (gsub (".*_", "MusDissSim_", DEfile), MusDissSim)
  rm (DEfile, DEtest, DEtestSig, MusDissSim)
}
```


```{r DEmusclePlots1, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 5, fig.width = 5}
for (MusDissSimfile in ls (pattern = "MusDissSim_")){
  MusDissSim <- get (MusDissSimfile)
  PLOT <- gplots::heatmap.2 (MusDissSim, col = colorRampPalette (c ("white", "black")), 
            trace = "none", margins = c (7, 7), breaks = 100, main = gsub (".*_", "", MusDissSimfile))
  PLOT
  rm (MusDissSimfile, MusDissSim)
}
DEallsetsMusBatch <- matrix (0, nrow = length (unique (c (rownames (DEtestSigBatch), rownames (DEtestSig_Raw) , rownames (DEtestSig_Batch), rownames (DEtestSig_BatchMus)))), ncol = 4)
rownames (DEallsetsMusBatch) <- unique (c (rownames (DEtestSigBatch), rownames (DEtestSig_Raw) , rownames (DEtestSig_Batch), rownames (DEtestSig_BatchMus)))
colnames (DEallsetsMusBatch) <- c ("SigBatch", "DESigMus_Raw", "DESigMus_Batch", "DESigMus_BatchMus")
DEallsetsMusBatch[, "SigBatch"] <- ifelse (rownames (DEallsetsMusBatch) %in%
                                             rownames (DEtestSigBatch), 1, 0)
DEallsetsMusBatch[, "DESigMus_Raw"] <- ifelse (rownames (DEallsetsMusBatch) %in% 
                                                 rownames (DEtestSig_Raw), 1, 0)
DEallsetsMusBatch[, "DESigMus_Batch"] <- ifelse (rownames (DEallsetsMusBatch) %in% 
                                                   rownames (DEtestSig_Batch), 1, 0)
DEallsetsMusBatch[, "DESigMus_BatchMus"] <- ifelse (rownames (DEallsetsMusBatch) %in%
                                                      rownames (DEtestSig_BatchMus), 1, 0)
pheatmap::pheatmap (DEallsetsMusBatch, color = c("grey", "darkblue"),
                      legend = F, show_rownames = F, treeheight_row = 0)

```

The GAL is the most different muscle.
The VAM, REF and VAL are similar and clustered together and GRA, SEM and SED are also more similar as they also cluster together.

```{r DEsummary1, echo = FALSE, warning = FALSE, message = FALSE}
for (DEfile in ls (pattern = "DEtest_")){
  print (DEfile)
  DEtest <- get (DEfile)
  DEtestSig <- get (gsub ("_", "Sig_", DEfile))
  # Summary of DE results
  contrasts <- gsub (" p.value", "", colnames(DEtest)[ grep ("p.value" ,colnames (DEtest))])
  for (i in contrasts) {
    M1 <- gsub (" -.*", "", i)
    M2 <- gsub (".*- ", "", i)
    LogCol <- colnames (DEtest)[ grep (paste0 ("Log2FC", "_", M1, "_", M2),colnames (DEtest))]
    PvalCol <- colnames (DEtest)[ grep (paste0 (i, " p.value"),colnames (DEtest))]
    HigherInM1 <- rownames (DEtestSig [DEtestSig [PvalCol] < 0.05 & DEtestSig [LogCol] > 0,])
    HigherInM2 <- rownames (DEtestSig [DEtestSig [PvalCol] < 0.05 & DEtestSig [LogCol] < 0,])
    assign(paste0 ("HigherIn", M1, "than", M2, gsub (".*_", "_", DEfile)), HigherInM1)
    assign(paste0 ("HigherIn", M2, "than", M1, gsub (".*_", "_", DEfile)), HigherInM2)
    rm (HigherInM1, HigherInM2)
  }

  summaryDETbale <- data.frame (Comparison = ls (pattern = paste0 ("HigherIn.*",  
                                                                   gsub (".*_", "_", DEfile), "$")), 
                                NoGenes = unlist(lapply (ls (pattern = paste0 ("HigherIn.*", 
                                                                              gsub (".*_", "_", DEfile), "$")),
                                                        function(x) length(get(x)))))
  assign (gsub (".*_", "summaryDETable_", DEfile), summaryDETbale)
  print (knitr::kable (head (summaryDETbale)))
  rm (summaryDETbale, DEtest, DEtestSig, DEfile)
}
```

Below plot visualizes the overlap of higher expressed genes than any other muscles.
There are many genes that are only higher in GAL than one or more other muscles
```{r musclePlots2, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 6, fig.width = 6}
for (DEfile in ls (pattern = "DEtest_")){
  print (DEfile)
  # Genes higher expressed in each muscle in comparison with any other muscle
  for (Mus in names (col.muscle)) {
    GenesHighInMus <- unique (unlist (lapply (ls (pattern = paste0 ("HigherIn", Mus, ".*", gsub (".*_", "_", DEfile), "$")), get)))
    assign (paste0 ("GenesHighIn", Mus, gsub (".*_", "_", DEfile)), GenesHighInMus)
    rm (GenesHighInMus)
  }
  
  # Genes that are higher at least in one muscle
  GenesHigh <- unique (unlist (lapply (ls (pattern = paste0 ("GenesHighIn.*", gsub (".*_", "_", DEfile), "$")), get)))
  AllSets <- matrix (0, ncol = 7, nrow = length (GenesHigh))
  rownames (AllSets) <- GenesHigh
  colnames (AllSets) <- ls (pattern = paste0 ("GenesHighIn.*", gsub (".*_", "_", DEfile), "$"))
  for (set in ls (pattern =  paste0 ("GenesHighIn.*", gsub (".*_", "_", DEfile), "$"))){
    AllSets [,set] <- ifelse (rownames(AllSets) %in% get(set), 1, 0)
  }
  assign (paste0 ("AllSets", gsub (".*_", "_", DEfile)), AllSets)
  pheatmap::pheatmap (AllSets, color = c("grey", "darkblue"),
                      legend = F, show_rownames = F, treeheight_row = 0)
  rm (AllSets, DEfile)
}
```
  
The genes that are only high in one muscle than any other muscle could be considered as the muscle specific markers?

```{r DEsummary2, echo = FALSE, warning = FALSE, message = FALSE}
for (AllSetsfile in ls (pattern = "AllSets_")){
  print (AllSetsfile)
  DEtestSig <- get (gsub ("AllSets_", "DEtestSig_", AllSetsfile)) 
  AllSets <- get (AllSetsfile)
  for (Mus in names (col.muscle)){
    COL <- paste0 ("GenesHighIn", Mus, gsub (".*_", "_", AllSetsfile))
    OnlyInMus <- as.data.frame (AllSets) %>% 
      tibble::rownames_to_column() %>%
      filter ((!!sym(COL)) == 1) %>%
      filter (rowSums(.[2:8]==1) == 1) %>% .$rowname
    assign (paste0 ("OnlyIn", Mus, gsub (".*_", "_", AllSetsfile)) , OnlyInMus)
    ls (pattern = paste0 ("HigherIn", Mus, gsub (".*_", ".*_", AllSetsfile), "$"))
  OnlyInMusthanAllOthers <- Reduce (intersect, lapply (ls (pattern = paste0 ("HigherIn", Mus, gsub (".*_", ".*_", AllSetsfile), "$")), get))
  OnlyInMusthanAllOthers <- OnlyInMusthanAllOthers [order (match (OnlyInMusthanAllOthers, rownames(DEtestSig)))]         
  assign (paste0 ("OnlyIn", Mus, "thanAllOthers", gsub ("AllSets_", "_", AllSetsfile)), OnlyInMusthanAllOthers)
  rm (OnlyInMus, OnlyInMusthanAllOthers)
  }
  DEsummary <- t (data.frame( `Only higher in one muscle than at least one other muscle` =  unlist (c (lapply (lapply (ls (pattern = paste0 ("OnlyIn.*", gsub ("AllSets_", "_", AllSetsfile), "$")) [! ls (pattern = paste0 ("OnlyIn.*", gsub ("AllSets_", "_", AllSetsfile), "$")) %in% ls(pattern = paste0("OnlyIn...th.*", gsub ("AllSets_", "_", AllSetsfile), "$"))], get), length))),
                              `Only higher in one muscle than all other muscles` = unlist (c (lapply (lapply (ls (pattern = paste0 ("OnlyIn...th.*", gsub ("AllSets_", "_", AllSetsfile), "$")), get), length)))))
  colnames (DEsummary) <- names (col.muscle)
  print (kable (DEsummary, align = "c"))
  assign (gsub ("AllSets_", "DEsummary_", AllSetsfile), DEsummary)
  rm (DEsummary, DEtestSig, AllSetsfile, AllSets)
}

library(VennDiagram)
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")

# Chart
for (Mus in names (col.muscle)){
  OnlyIn_BatchMus <- get (paste0 ("OnlyIn", Mus, "_BatchMus"))
  OnlyIn_Batch <- get (paste0 ("OnlyIn", Mus, "_Batch"))
  OnlyIn_Raw <- get (paste0 ("OnlyIn", Mus, "_Raw"))
  venn.diagram (
    x = list (OnlyIn_BatchMus, OnlyIn_Batch, OnlyIn_Raw),
    category.names = c("BatchMus" , "Batch" , "Raw"),
    filename = paste0 ("figures/VennDiagramDEoverlapOfDataSets/" ,"HigherIn", Mus, ".png"),
    output=TRUE,
    # Output features
    imagetype="png" , height = 480 , width = 480 , resolution = 300, compression = "lzw", 
    # Circles
    lwd = 2, lty = 'blank', fill = myCol, 
    # Numbers
    cex = .6, fontface = "bold", fontfamily = "sans", 
    # Set names
    cat.cex = 0.6, cat.fontface = "bold", cat.default.pos = "outer", cat.pos = c(-27, 27, 135),
    cat.dist = c(0.055, 0.055, 0.085), cat.fontfamily = "sans", rotation = 1
  )
  rm (OnlyIn_BatchMus, OnlyIn_Batch, OnlyIn_Raw)
  OnlyInThanAll_BatchMus <- get (paste0 ("OnlyIn", Mus, "thanAllOthers_BatchMus"))
  OnlyInThanAll_Batch <- get (paste0 ("OnlyIn", Mus, "thanAllOthers_Batch"))
  OnlyInThanAll_Raw <- get (paste0 ("OnlyIn", Mus, "thanAllOthers_Raw"))
  venn.diagram (
    x = list (OnlyInThanAll_BatchMus, OnlyInThanAll_Batch, OnlyInThanAll_Raw),
    category.names = c("BatchMus" , "Batch" , "Raw"),
    filename = paste0 ("figures/VennDiagramDEoverlapOfDataSets/" ,"HigherIn", Mus, "thanAllOtherMuscles.png"),
    output=TRUE,
    # Output features
    imagetype="png" , height = 480 , width = 480 , resolution = 300, compression = "lzw", 
    # Circles
    lwd = 2, lty = 'blank', fill = myCol, 
    # Numbers
    cex = .6, fontface = "bold", fontfamily = "sans", 
    # Set names
    cat.cex = 0.6, cat.fontface = "bold", cat.default.pos = "outer",
    cat.dist = c(0.055, 0.055, 0.085), cat.fontfamily = "sans", rotation = 1
  )
  rm (OnlyInThanAll_BatchMus, OnlyInThanAll_Batch, OnlyInThanAll_Raw)
}

```

### Heatmap of log-CPM values for 1000 top differentially expressed genes
```{r musclePlots4, echo = FALSE, warning = FALSE, message = FALSE, fig.height = 14, fig.width = 7}
for (DEtestSigfile in ls (pattern = "DEtestSig_")){
  print (DEtestSigfile)
  DEtestSig <- get (DEtestSigfile)
  # Heatmap for top DE genes 
  # Get the genes for the top 1000 most DE genes
  HighDE <- rownames (DEtestSig [order(DEtestSig$`FDR(>F)`),])[1:1000]
  assign (gsub ("DEtestSig_", "HighDE_", DEtestSigfile), HighDE)
  # Subset matrix
  lcpm <- cpm (get (gsub ("DEtestSig_", "D_", DEtestSigfile)), log = T)
  hmDat <- lcpm [HighDE, ]
  # Set the colors
  myCols <- cbind (BatchColPlot, MusColPlot, IndColPlot)
  colnames (myCols) <- c ("Batch", "Muscle", "Individual")
  # Plot the heatmap
  heatmap3::heatmap3 (hmDat, ColSideColors = myCols,
              labCol = gsub ("_Sam.*", "", colnames (hmDat)), 
              main = paste ("Top 1000 DE genes", DEtestSigfile), 
              labRow = FALSE, scale = "row",
              col =   gplots::colorpanel (1000, "blue", "white", "red"),
              margins = c(5,1), Rowv = T, showRowDendro = F)
  rm (HighDE)
}
table(HighDE_Batch %in% HighDE_BatchMus)
table(HighDE_Batch %in% HighDE_Raw)
table(HighDE_Raw %in% HighDE_BatchMus)

```

Since there are many GAL muscles in the batch 5 (resequencing batch), we had this concern that including the batch as a covariate in the DE model would not entirely correct for the batch. So as you see in the above figures we generated 2 corrected count data (1) only correcting batch (2) correcting for batch and including the muscle as a covariate to the model.
DE analysis was done in 3 different ways:

(DE1) not corrected data so the batch is included as a covariate
- lmer (values ~ muscle + batch + (1|individual), weights = weights, data = DE.Data) 
(DE2 and DE3) 2 corrected count datasets without including the batch in the model
- lmerTest::lmer (values ~ muscle + (1|individual), weights = weights, data = DE.Data)


The ComBat-Seq considering the muscle as group (DE3) does what we need.
in the DE2 since the batch correction is done without considering the muscle as a trait to preserve we are losing some biology because there are 8 GAL muscles in the batch 5 (the problematic batch).
There is a high overlap between the DE1 and DE3 however since the plot for 1000 top DE genes shows less clustering based on batch so we decided to continue with this dataset.

```{r SaveSelectedDataset, echo = FALSE, message = FALSE, warning = FALSE}
if (! file.exists ("Outputs/BatchMusCorrectedAllData.RData")) {
  ObjectNames <- noquote (ls (pattern = "BatchMus"))
  save (list = ObjectNames, file = "Outputs/BatchMusCorrectedAllData.RData")
}
```

### Enrichment analysis
```{r EnrichmentAnalysis, message = FALSE, warning = FALSE}
if (!file.exists ("Outputs/DE/EnrichmentResultsDEGenes.RData")) {
  rm (list = ls ())
  Data <- new.env(); load ("Outputs/BatchMusCorrectedAllData.RData", envir = Data )
  v_BatchMus <- Data$v_BatchMus
  
  DEEnrichment <- c ()
  for (GeneList in ls (envir = Data, pattern = "OnlyIn|HigherIn|GenesHighIn")) {
    print (GeneList)
    List <- get (GeneList, envir = Data)
    if (length (List) != 0) {
      GeneList <- gsub ("_.*", "", GeneList)
      BG <- as.vector (rownames (v_BatchMus))
      GP <- gprofiler2::gost (List, organism = 'hsapiens', evcodes = TRUE,
                              custom_bg = BG, correction_method = "fdr") 
      GP <- GP$result
      if (! is.null (GP)) {
        DEEnrichment [[GeneList]] <- GP
      }
    }
  }
  # The final table
  DEEnrichment <- do.call (rbind, DEEnrichment)
  DEEnrichment[c ("query", "significant", "precision", "recall")] <- NULL
  DEEnrichment <- DEEnrichment %>% tibble::rownames_to_column("List") %>% mutate (parents = sapply (parents , toString))
  
  ## Gene mapping
  Gene_Symbols <- c ()
  for (i in 1:nrow (DEEnrichment)) {
    ENS <- DEEnrichment [i, "intersection"]
    ENS <- unique (unlist (strsplit (as.character (ENS), split = ",")))
    Gene_Symbol <- c ()
    for (j in ENS) {
      Gene_Symbol [j] <- v_BatchMus$genes [v_BatchMus$genes$ensembl_gene_id == j, "hgnc_symbol"]
    }
    
    Gene_Symbols [[i]] <- paste (Gene_Symbol, collapse = ",")
  }
  
  Gene_Symbols <- do.call (rbind, Gene_Symbols)
  DEEnrichment <- cbind (DEEnrichment, Gene_Symbols)
  save (DEEnrichment, file = "Outputs/DE/EnrichmentResultsDEGenes.RData")
  
  DEEnrichment_Filtered <- DEEnrichment  %>% 
    filter (intersection_size > 3) %>% filter (term_size < 1000) %>%
    mutate (List = gsub ("\\..*", "", List)) %>% 
    mutate (evidence_codes = NULL)
  
  write.csv (DEEnrichment_Filtered, "Outputs/DE/EnrichmentResultsDEGenesIntersect3.csv", row.names = F)
}
```
